Changes to the NH fix enabling Cauchy stress control (Cauhchystat) du…

ProfessorMiller opened the pull request:
…e to Miller, Tadmor, Gibson, Bernstein and Pavia, J Chem Phys, 144, 184107 (2016).

## Purpose

This is an extension of the "fix npt" command that allows for direct control of the true (Cauchy) stress instead of the second Piola-Kirchhoff stress (PK stress).  The current Parrinello-Rahman algorithm in the stable release of "fix npt" controls the PK stress.  If the simulation box shape changes significantly during the run, the PK stress can be considerably different than the set stress values.  This fix implements the so-called "Cauchystat" of the paper cited above to control the true stress even when the box shape changes.

## Author(s)

Credit for the Cauchystat theory should go to the cited paper.  Authors of the Lammps implementation deposited here are:

Ronald E Miller, Carleton University
Fabio Pavia, formerly of EPFL and
Subrahmanyam Pattamatta, University of Minnesota

## Backward Compatibility

There is no problem with backward compatibility.  If the keyword "cauchystat" does not appear on the fix npt command line, everything behaves as before.

## Implementation Notes

Changes are localized to fix_nh.cpp and fix_nh.h.  Correctness was verified using the examples provided in examples/cauchystat included with this pull request.  These examples reproduce results published in Miller et al (2016).

## Post Submission Checklist

_Please check the fields below as they are completed_
- [ ] The feature or features in this pull request is complete
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] One or more example input decks are included
- [x] The source code follows the LAMMPS formatting guidelines

## Further Information, Files, and Links

This contribution is as per email exchanges with Aidan Thompson, and incorporate all of his suggestions.

akohlmey commented:
@sjplimp @athomps when polishing this contribution, it is probably going to be convenient to also review and possibly include the code from PR #1026 and apply equivalent polishing as well.

ProfessorMiller commented:
I've updated the pull request as per @athomps request.   I moved as much of the Cauchystat-related code out of the main body of fix_nh so it is much cleaner now.  I also left only a single example that runs in about 30 secs.  I tried to clean up the whitespace and indentation, but my cpp skills and knowledge of lammps conventions isn't great.  Let me know if I can do more.

akohlmey commented:
@athomps please select "Approve" instead of "Comment" for your review.

akohlmey commented on src/fix_nh.h:
```diff
@@ -139,6 +139,27 @@ class FixNH : public Fix {
+  double invH0[3][3];            //inverse of H0;
+  double CSvol0;                 //volume of undeformed cell
+  double setPK[3][3];            //current set values of the PK stress (this is modified until the cauchy stress converges)
+  static double setPKinit[6];    // initialization value of setPK for continuation runs
```
Please explain why `setPK[][]`, `restartPK` and `restart_stored` have to be static class members. It looks to me, as if the `static` qualifier can be removed, if those variables are initialized in the constructor.

ProfessorMiller replied:
setPK is not static, I think you mean setPKinit and the others (restartPK and restart_stored).  These three are static because they need to be kept up-to-date at the end of a NPT run so that they are known at the start of any subsequent NPT runs.  If there is a cleaner cpp way to do this, please advise - I am a cpp amateur.

akohlmey commented:
On Sun, Sep 30, 2018 at 9:55 AM ProfessorMiller <notifications@github.com>
wrote:

> *@ProfessorMiller* commented on this pull request.
> ------------------------------
>
> In src/fix_nh.h
> <https://github.com/lammps/lammps/pull/1096#discussion_r221460691>:
>
> > @@ -139,6 +139,27 @@ class FixNH : public Fix {
>    double compute_strain_energy();
>    void compute_press_target();
>    void nh_omega_dot();
> +
> +  // Implementation of CauchyStat
> +  double H0[3][3];               //shape matrix for the undeformed cell
> +  double h_old[6];               //previous time step shape matrix for the undeformed cell
> +  double invH0[3][3];            //inverse of H0;
> +  double CSvol0;                 //volume of undeformed cell
> +  double setPK[3][3];            //current set values of the PK stress (this is modified until the cauchy stress converges)
> +  static double setPKinit[6];    // initialization value of setPK for continuation runs
>
> setPK is not static, I think you mean setPKinit and the others (restartPK
> and restart_stored). These three are static because they need to be kept
> up-to-date at the end of a NPT run so that they are known at the start of
> any subsequent NPT runs. If there is a cleaner cpp way to do this, please
> advise - I am a cpp amateur.
>

yes, i think you are misinterpreting what the static keyword does. that can
easily happen, as it has confusingly many meanings in c++. the value of all
class members is maintained for the lifetime of the instance of the class
(from fix command to unfix command). there is a separate copy for each
class instance (be it a LAMMPS instance or one of multiple fix command
instances that are active at the same time). static class members on the
other hand exist only once, just like a global variable. this is not the
desired behavior in general, even though it is mostly compatible with
typical uses.

the behavior you describe is possible by simply removing the "static"
qualifier and initializing those class members in the constructor (if - in
contrast - they would need to be initialized fresh for each MD run, you
would initialize them in the setup method instead).

best you try it out and see if what i am saying is correct.

axel.



> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lammps/lammps/pull/1096#discussion_r221460691>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARp4yISokWkQALgB50ljHb2ZvYKQtfcks5ugM1hgaJpZM4WZpL_>
> .
>


-- 
Dr. Axel Kohlmeyer  akohlmey@gmail.com  http://goo.gl/1wk0
College of Science & Technology, Temple University, Philadelphia PA, USA
International Centre for Theoretical Physics, Trieste. Italy.

ProfessorMiller commented:
Hi Axel,

I'll have a look at what you say.  But one more question, to make sure I understand.  Consider the bit of a lammps input script below.  What is does is:
- equilibrate at zero stress (first fix/run combination)
- ramp stress down from 0 to -10000.0 (second fix/run combination)
- ramp stress back from -10000.0 to 0 (third fix/run combination)
Note the deliberately inconsistent use of unfix

> fix 1 all npt temp 600.0 600.0 1.0 couple none cauchystat 0.001 no
> run 10000
> unfix 1
> fix 1 all npt temp 600.0 600.0 1.0 xy 0.0 -10000.0 0.1 couple none cauchystat 0.001 yes
> run 10000
> fix 1 all npt temp 600.0 600.0 1.0 xy -10000.0 0.0 0.1 couple none cauchystat 0.001 yes
> run 10000

I need the values of those variables, stored at the end of the first instance of fix 1, to be available to the algorithm at the start of the second instance.  Then I need the variables from the end of the second instance to be available at the start of the third instance.  

This behaves correctly with the static keywords in place.  If I understand your comment though, it would not work for the first transition between fixes, because the data would be lost at the unfix.  It would be ok, I think, in the second transition, but users may not appreciate that in order to create a multi-step loading program requires them to use and modify a single fix with no "unfix" in between.  

Because of this, I think it may really be that I need a global variable.

akohlmey commented:
@ProfessorMiller i disagree with your assessment of the situation. the general behavior of fix styles in LAMMPS is, that if you do an `unfix`, you have a **complete** reset of all settings, as it deletes the instance of the fix class and creates a fresh one. Expecting, that data is carried over such a case is against the conventions in LAMMPS and can cause all kinds of problems. Instead the behavior that in order to change settings on a fix, you re-issue the fix command with the same fix style and fix ID, is the convention in LAMMPS. for this you do not need to have static class members. but the static class members can do harm, too. consider somebody is running an input with multiple separate simulations, by using the `clear` command and then setting up a new system. with using static class members, you would apply the information from the previous simulation, which is clearly **not** what you want. so the only way to reliably get a reset is that you allow the reset via unfix followed by fix and avoid the reset by re-issuing the fix command with the same style and different parameters.

after this discussion and the additional information you provided i now believe more than before, that the use of `static` is not resulting in a behavior that is consistent with the rest of LAMMPS. 

this prompts another question. how would you be handling the case of somebody using the `cauchystat` flag for the first run, then (without unfix) re-issues the fix **without** that flag, and then does another run **with** it again. both times without unfix. in this case, i believe the consistent behavior would be that the state is not carried over, even without using unfix. since the internal data structures are not going to be updated.

akohlmey commented:
@ProfessorMiller one more comment. if people contributing code believe that the typical (and expected at least by the developers) behavior of LAMMPS may confuse users, we usually encourage to place a `NOTE:` paragraph at a prominent place into the documentation for the contributed code explaining how to achieve the desired behavior.

ProfessorMiller commented:
Hi Axel,

Ah, I see - this is showing my ignorance of the lammps structure and approach, but I am learning.  I agree that given the lammps conventions, they should not be static.  

Incidentally, the examples you gave of misuse of the cauchystat command is defended against, to some extent, by the "continue" flag.  The user would be doing something meaningless, as you correctly point out, in both those examples, but setting continue to "no" would cause cauchystat to ignore the stored values and start fresh.

In any event, I now see your point and understand.  If I assume unfix is not used in between, I think I can get the correct behaviour.  I also just saw your other note, and I was thinking the same thing, that I should include a note in the documentation to this effect.  Will do.

akohlmey commented:
@ProfessorMiller this is now the updated version rebased to the current patch release with the changes also in #1171 
watch out when updating. this will become a "forced update" and may cause problems, if you have committed additional local changes.

akohlmey commented:
@sjplimp please have a look at the use of fix STORE to retain data between re-issued variants of the fix command.

akohlmey commented:
@ProfessorMiller i have not been able to do any programming on this, but there is one nagging thought, that i cannot shake off. Can you please explain, why you need to re-issue the thermostat fix command for a second run (which is causing all the issues with carrying over data), instead of just leaving the fix command untouched and continue using the data that is already in the various data structures? i cannot find any part of the code, where the fix re-issue would be required, as the data that is carried over in fix STORE remains available, and initialization from scratch can just be skipped. but i may be overlooking something...

ProfessorMiller commented:
@akohlmey:  I suppose that there might be a way to do it differently.  My understanding of the npt fix is that it is limited to applying either a constant stress or a stress that linearly varies from Pstart to Pstop. 

 If one wants to do a more complicated loading program, one uses multiple fix-run pairs.  Say I want to go from 0 to 1000bars and then back down from 1000bars to 0 - I would use 2 fix-run pairs for that.  Something like:

fix 1 all npt temp 600.0 600.0 1.0 z  0.0 1000.0 0.1 couple none cauchystat 0.001 no
run 100000
fix 1 all npt temp 600.0 600.0 1.0 z 1000.0 0.0 0.1 couple none cauchystat 0.001 yes
run 100000

The second fix for the decreasing stress could start without the stored information from the previous run, should generally "work" in that it will be stable, but it is likely to introduce some extra movement in the stress values at the start of the run.  This is because the algorithm has to guess where to start the applied stress levels.  If it doesn't know anything about the previous set values, it initializes to the current Cauchy stress values, which (because of natural stress fluctuations in the NPT ensemble) are somewhat different than the mean value to which the previous run converged.  It will still get to the right place, but might take 1000 time steps (say) to re-establish equilibrium.  Passing the data between fixes allows the second run to start seamlessly from the first.

If there is another way to run a load program like this example above that avoids the re-initialization of the fix, that would be fine.

akohlmey commented:
@ProfessorMiller thanks for the insight this is helpful to understand the needs. We have discussed this pull request and the associated changes among the core LAMMPS developers today, and in that context we discovered, that there are some pending changes coming down the line that are addressing some known issues with anisotropic cell changes. We have determined, that those should have priority over your changes and thus we would like to propose a different way to get your feature integrated:
- create a new package/folder `USER-CAUCHY` or `USER-CAUCHYSTAT`
- put a copy of your modified `fix_nh.cpp/fix_nh.h` into that folder and rename it to `fix_cauchy.cpp/.h`
- create derived classes that would provide `fix npt/cauchy` (`fix_npt_cauchy.cpp/.h`) and `fix nph/cauchy` (`fix_nph_cauchy.cpp/.h`) and whatever other variant you deem worth providing
- remove all code and flag/variables not necessary for "cauchystat" mode for easier maintenance
- document those styles in a separate file (`doc/src/fix_cauchy.txt`)

This will result in some code duplication, but it avoids making the code in fix_nh.cpp more complex and it would also liberate you from having to provide binary restart compatibility (even though it might be beneficial), as these fixes would now be maintained and documented separately.

Adding your feature as a separate package would allow us to merge it sooner and the feature would be available to people interested in it faster, without risking to compromise the complex and thus fragile code in fix_nh.cpp and the many other classes that derive from it.

Does this sound reasonable and acceptable?

ProfessorMiller commented:
@akohlmey, that sounds fine with me.

akohlmey commented:
ok.

if you want, i can do the basic changes (copying files, renaming classes, setting up installation/compilation) in this branch for you with just one example "production" fix (say, `fix npt/cauchy`), but then leave the trimming of the code, additional derived classes etc. and updating the documentation to you. that would get you started without having to worry much about how to integrate a new package (and i often have to do this process when merging contributed changes anyway, so it is familiar to me).

or you could close this pull request and start over yourself if you prefer.

please let me know.

ProfessorMiller commented:
@akohlmey, if you can do the basic changes that would be a big help.  I'll take it from there.  Thanks.

akohlmey commented:
@ProfessorMiller here you go. This is just copying the code and docs to new files/folders, renaming classes, int and removing the changes to `fix_nh.cpp/.h/.txt`

I would remove the "cauchystat" keyword and instead make "alpha" and "continue" a keyword each.
and then there is probably a chunk of code that can be removed from the parser and the implementation, so that you have only features included that are used.

ProfessorMiller commented:
@akohlmey I think I have correctly committed a clean working version.  Please have a look.  I edited the fix_nh.txt doc to match the current state.  Should was also add the USER-CAUCHY package here:
https://lammps.sandia.gov/doc/Packages_details.html
?

akohlmey commented:
This is now updated to the latest master, documentation has been adapted for our new scheme, the code has been merged into a single pair of files and thus moved to the USER-MISC package as customary for such cases. A lot of inconsistencies of error messages and warnings have been resolved and the fix is referred to as npt/cauchy consistently. Also some "Fortran style" array initializers were replaced with C++ code. Pending any further issues that will be found by our testing, this should be ready for inclusion now.
Sorry for the long delay, we have accumulated a bit of a backlog over the last couple of years and pending changes, that required some work, like this one and a few others, were thus sinking to the bottom of the pile. We are making a conscious effort to get rid of this backlog currently.

ProfessorMiller commented:
Thanks Axel!  I appreciate all your effort on this.
